core-zipline/src/androidMain/kotlin/io/github/droidkaigi/confsched2022/zipline/SessionsZipline.kt

```diff
@@ -62,43 +63,50 @@ class SessionsZipline @Inject constructor(
         nowEpochMs = { System.currentTimeMillis() },
     )
 
-    fun timetableModifier(
-        coroutineScope: CoroutineScope,
-    ): MutableStateFlow<suspend (DroidKaigiSchedule) -> DroidKaigiSchedule> {
-        val androidScheduleModifier = AndroidScheduleModifier()
-        val defaultModifier: suspend (DroidKaigiSchedule) -> DroidKaigiSchedule = { timetable ->
-            androidScheduleModifier.modify(timetable)
+    // Limit the write access only on the Zipline thread
+    private val mutex = Mutex()
+    private var cachedScheduleModifier: ScheduleModifier? = null
+    private suspend fun takeOrGetScheduleModifier(): ScheduleModifier = withContext(dispatcher) {
+        mutex.withLock {
+            val cached = cachedScheduleModifier
+            if (cached != null) {
+                return@withLock cached
+            }
+            val loadedZipline = ziplineLoader.loadOnce(
+                applicationName = "timeline",
+                manifestUrl = manifestUrl,
+                initializer = { },
+            )
+            val taken = loadedZipline.zipline.take<ScheduleModifier>("sessionsModifier")
+            cachedScheduleModifier = taken
+            taken
         }
-        val modifierStateFlow = MutableStateFlow(defaultModifier)
+    }
 
-        coroutineScope.launch(dispatcher) {
-            var zipline: Zipline? = null
+    @OptIn(ExperimentalTime::class) // measureTimedValue
+    override fun timetableModifier(): Flow<suspend (DroidKaigiSchedule) -> DroidKaigiSchedule> {
+        return channelFlow {
+            // The JS modifier takes about 300 ms to execute.
```

Comment:
The JS modifier -> Loading of JS modifier
This can lead misunderstanding👀

Reply:
Improved this comment.
Please check [05cfe3b](https://github.com/DroidKaigi/conference-app-2022/pull/505/commits/05cfe3b862d501b12528fc3119ee0e6a46139b4f).